slides/models-communication.md

```diff
@@ -0,0 +1,135 @@
+  - e.g., a web server
+    * Can just provide data (as in static web content)
+	* Or can do work on behalf of client (as in personalized or dynamic data generation)
+  - in multi-tier systems, servers can themselves be clients of other servers.
+
+The client is typically less interested in the location or
+other physical characteristics of the server and more so in the service that it
+provides.
+
+* Client initiates work by naming the server
+* That server name must be _resolved_
+  - Possibly to a job on a specific backend machine known to the client
+  - Or perhaps to a collection of machines that perform this service
+	* Load-balanced within a data center to provide scalable performance
+	* And/or distributed across multiple locations to provide fast communication
+* Client makes a _request_ to the server and the server provides an appropriate _response_
+* Requests are typically considered a single atomic unit of work 
+  - Doesn't form a long-term association between any particular client
+    or server.
+
+Because of the dichotomy between clients and servers, these often have very
+different hardware characteristics and use different application-design models
+internally.
+
+* Servers often optimized for high scalability and throughput
+  - Don't want them bogged down by any one request.
+
+## Streaming data
+
+In client-server communication, a client requests information from a server but
+does not set up a longer-term association. In some types of workflows, though,
+different subtasks have a long-term predictable pattern of communication
+flow. 
+
+* Stream of data flow between associated subtasks
+  - A stream is a reliable full-duplex mode of communication, with
+    messages on each side of the stream being delivered to the other
+    side in order. 
+  - Can be more efficient than client-server since the startup
+    overhead of the communication only takes place at the beginning of
+	a long-lived stream of messages rather than on each new message
+* No hierarchy or dichotomy required between different subtasks
+  - All can be equal logical participants
+  - The data flowing in these streams enables them to coordinate and
+    perform appropriate work
+* Can be merged with client-server model
+  - The client initiates the stream, but after that, both sides are equal
+    participants (bidirectional/bidi streaming)
+  - or client-only streaming: the client initiates the stream and sends many 
+    mesages, after which the server sends a single response
+  - or server-only streaming: the client sends a request and the server
+	responds with a stream of messages
+* Disadvantage
+  - Long-term association may make it difficult to dynamically change the
+    back-end server associated with a particular stream of communication
+	* Limits load-balance opportunities to once per stream instead of once
+	  per mesage
+
+## Controller-worker
+
+* Another model with a dichotomy of subtask classes
+  - controller
```

Comment:
I'd include a one-liner describing what a controller/worker is. Otherwise just introducing the name accomplishes little without something to associate it with right away.